### ISSUE TITLE
Does state name passing with parameter on event make sense? 

### ISSUE THREAD

**Issue by youroff** [aasm/aasm#324]
What's the reason in passing desired state name when calling the event with parameters? As far as I understand the idea behind finite state machines, event should unambiguously set what state it will end at. We have guards to implement conditional transitions etc, so why there's a first parameter? I think that api would be cleaner if we could pass arbitrary parameters for callbacks.

Is there a use-case I don't see right now?

**Comment by alto** [comment]
Hi @youroff,

_AASM_ supports not only deterministic but also non-deterministic state machines. So sometimes the next target might be ambiguous and _AASM_ therefore offers a way to define your requested target state (when firing the event). Since version `4.0.0` this parameter is not mandatory anymore (I will update the README on this). Feel free to omit it.

Does that make things clear?

**Comment by youroff** [comment]
Yeah, after some experiments I found that it works without it, thank you! However, it doesn't look like non-deterministic state machine, as it makes exact the same transition based on the input. If no desired state passed, then first matching by guards, if it's passed, then transition to that state. So it just makes reasoning about the events harder.

**Comment by hugueslamy** [comment]
I'll just piggyback on this to explain the case why I would need the destination state on an event. Now I don't know if this is bad design or something that was making sense before and most people just not use it enough.

To illustrate this, let explain with a food distributor which use coins.
Normally, you would have a **wait** state with a event coin_inserted (coin_value_params) that loop back to wait until the event **buy** is triggered. This would be the oversimplified flow.

Now, suppose I have a third party that send me event via a single patch request with the parameters state and value. When adding the coin I receive PUT /transactions/:id {state: **wait**, value: 5}. For this system, the transition already happened and I get the end state. Now my system must follow what is provided to me. On the later call I got a PUT /transactions/:id {state: **buy**, value: nil}. 

My choices are now the following:

1. Direct assignment of the state. Now the guard and events are not called so this is a no-no.

1. Create 2 events - Insert coins and Buy and have the controller select the right one with the help of an ugly case.

1. Create only 1 event called human_action, with the end state and parameters as describe in the aasm documentation.

And now for the bug: When using the idea 3, if you put an invalid state in the first parameter, the system will just use the first possible transition. Looking at the _fire method in the event.rb file
```
# If to_state is not nil it either contains a potential
# to_state or an arg
unless to_state == nil
if !to_state.respond_to?(:to_sym) || !transitions.map(&:to).flatten.include?(to_state.to_sym)
   args.unshift(to_state)
   #to_state = nil
   end
end
```
If the to_state send is not in the possible list then add it to the arguments to be passed to the block.

This cause two issues:
If the state is invalid:
human_action(:blah) the first available transition based on the from state will be selected.
if the state a valid name but not possible from any state to this state, the first available transition based on the from state will be selected.
For me I think this break the state machine flow as the state and transition are not enforce anymore. 
On the first case, my expectation would be to receive a
`raise AASM::UndefinedState, "State :#{to_state} doesn't exist" if to.nil?`
On the second case I would expect an invalid transition.

I did a small correction on the event.rb file, in the _fire method. It was late last night, so I didn't test it much, and I'm pretty sure that will fail the test, but this is a starting point for you.

```
    # Execute if test == false, otherwise return true/false depending on whether it would fire
    def _fire(obj, options={}, to_state=nil, *args)
      to = obj.aasm.states.find {|s| s.name == to_state}
      raise AASM::UndefinedState, "State :#{to_state} doesn't exist" if to.nil?

      result = options[:test_only] ? false : nil
      if @transitions.map(&:from).any?
        transitions = @transitions.select { |t| t.from == obj.aasm(state_machine.name).current_state }
        return result if transitions.size == 0
      else
        transitions = @transitions
      end

      # If to_state is not nil it either contains a potential
      # to_state or an arg
      unless to_state == nil
        if !to_state.respond_to?(:to_sym) || !transitions.map(&:to).flatten.include?(to_state.to_sym)
          args.unshift(to_state)
          #to_state = nil
        end
      end

      transitions.each do |transition|
        next if to_state and !Array(transition.to).include?(to_state)
        if (options.key?(:may_fire) && Array(transition.to).include?(options[:may_fire])) ||
           (!options.key?(:may_fire) && transition.allowed?(obj, *args))
          result = to_state || Array(transition.to).first
          if options[:test_only]
            # result = true
          else
            transition.execute(obj, *args)
          end

          break
        end
      end
      result
    end
```

If you don't see the changes, on the second line, I start with a validation of the state. Some extra code should be put in to accept the nil and default to the next available transition. 
Toward the middle I commented the **to_state = nil**. This will force a validation check on the possible transition. Now I think that most people use the event function without the end state and most likely it should support this behaviour.

So I spent few hours on this. I'm guessing it is clear for me, but sometimes translating this to writing is another story and don't hesitate to drop me a note if you need more information or this is not clear enough.

**Comment by youroff** [comment]
I think it's better to rethink your state machine definition. Does coin deposit really moves it to another state? According to regular wending machines, it's just exactly the same when no coins inside or some amount that is enough (or not enough) for purchase. So when you try to buy something you either have enough coins or not. I'm not sure that state machine is needed here at all.

Could you post state machine definition to illustrate the workflow?

**Comment by hugueslamy** [comment]
Here's a simple state machine:
```
  aasm :column => 'status' do
    state :first
    state :second
    state :third

    event :run do
      transitions from: :first, to: :second
      transitions from: :second, to: :third
    end
  end
```
Now at the console let's assume we have an object job and the actual state is :first 
job.run -> job.status == :second -> This is correct.
job.reload
job.run(:second) -> job.status == :second -> This is correct
job.reload
job.run(:third) -> job.status == :second -> This is incorrect as this transition is not allowed. I was expecting InvalidTransition
job.reload
job.run(:blah) -> job.status == :second -> This is incorrect as this state is not defined anywhere. I was expecting UndefinedState

**Comment by pirj** [comment]
@youroff, real life example:

    aasm column: :state do
      state :unconfirmed, initial: true                                                                                                                                                                
      state :confirmed                                                                                                                                                                                 
      state :rejected                                                                                                                                                                                  
                                                                                                                                                                                                       
      event :confirm, after: :set_user_and_time do                                                                                                                                                     
        transitions from: :unconfirmed, to: :confirmed                                                                                                                                                 
      end                                                                                                                                                                                              
                                                                                                                                                                                                       
      event :reject, after: :set_user_and_time do                                                                                                                                                      
        transitions from: :unconfirmed, to: :rejected                                                                                                                                                  
      end
    end

    private

    def set_user_and_time user
      self.confirmed_or_rejected_by = user
      self.confirmed_or_rejected_at = Time.current
    end

In controller:

    order.confirm!(current_user)

Any better way to do so?

**Comment by youroff** [comment]
@pirj Not sure I understand the problem. Do you want to use one event and some business logic to decide to go whether to confirmed or rejected?

**Comment by youroff** [comment]
@hugueslamy I see the problem, but this problem exists only under assumption that sending of event + desired state is a good idea, but my point is opposite. In your case I would have separate events, like `to_second` and `to_third` and this would give you meaningful feedback, if you try to go to `third` from `first` calling `to_third`, for example. Actually this confusion is just another proof that desired state should be completely removed from event call.

**Comment by anujaware** [comment]
@hugueslamy :+1:  I think as per documentation if we pass ```nil ``` it will transition to the first end state defined for that event i.e. perfect but if parameter is present and its not the valid state it would be more correct to thrown an error. I tried by using bang to check if it throw error while saving the object but it doesn't.

**Comment by youroff** [comment]
@anujaware what if parameter passed is the piece of data that you want to see in event callback and it happened to be a string or symbol? It this case there's no way to distinguish what exactly developer meant.

**Comment by anujaware** [comment]
@youroff Can you please give a small example?

**Comment by youroff** [comment]
@anujaware Example of what exactly? Of passing a context with events? It was posted by @pirj couple of comments earlier: https://github.com/aasm/aasm/issues/324#issuecomment-204482526

**Comment by pirj** [comment]
@youroff If you think this is not the best way to pass context along with events, propose a better way, this is what I meant by "Any better way to do so?".
This is what your original question was, why these parameters are here.
Propose a better solution to my problem without the use of parameters. This might add to a chance for them to be deprecated in next major version.

**Comment by youroff** [comment]
@pirj Not really, my original question was about state passed as a first argument. I strongly believe that passing arbitrary arguments to event for later use in callbacks is the right way to go. But at the same time passing target state is a bad design that only brings confusion.

**Comment by anujaware** [comment]
@youroff @pirj Example is right. But I think we will need that end state as parameter to support non deterministic state machine but also need to find some other way to handle above case. Something like  either don't pass any argument or pass argument with end state (which could be nil).

**Comment by youroff** [comment]
@anujaware How exactly passing state name would make this state-machine support non-determinism? Doesn't passing explicit state make is deterministic? To be honest, I don't see real application for even emulated NFA, unless we have quantum computers or at least highly parallel platform, which is not Ruby of course. So when you pass state name with the event, you don't make it non-deterministic, you just decompose event into two or more different events (which is actually legit, as NFA can be converted into more complex DFA), but in real application it only produces a mess.

**Comment by hugueslamy** [comment]
@youroff If I look at UML, I think that you need different event to move to a different state. I'm encline to follow this simpler route when possible, but sometimes you're stuck with third party system that are not well design and passing the end state is a better solution to avoid spaghetti code.

So you end up with the following possibilities.
- event (without any argument) = nil end state so pick the first possible transition in the list
- event(:end_state, args) = Try to determine the last state from the first argument
- event(args) = This could be ambiguous with the previous call and could lead to problems.
- event(nil, args) = This fix the first argument to the end state and send everything as argument. This could break current implementation. This is the version that give the most flexibility.

My feeling on this is that the target state is sometimes very helpful even if not ideal. To support the actual code base, my feeling would be to add a configuration to correctly handle the behaviour of your code with the most sensible default to support only 1 & 3. On the other hand, if you support 1, 2 & 4 you must strictly enforce the transition and available state.

**Comment by anujaware** [comment]
@youroff Yeah you are right we can use different events. I am also not able to find real life scenario where we can not avoid passing end state and can not handle it with event.
@hugueslamy We can handle this by making end stare mandatory but I am also curious to know the exact example where we can't avoid passing end state.

**Comment by youroff** [comment]
@hugueslamy Maybe I miss something. What do you mean under 'third party service' here? Something that sends HTTP-requests to your api? But in this case you have a controller as a layer between that service and your state-machine, why not to deal with distribution there? I would vote for 1 & 3. For 4, that would be tedious to repeat nil everywhere, so I'm not sure. I think there is how it was in older versions.

@anujaware mandatory end state will make the use of guards with multiple transitions impossible, for me this is pretty important feature.

**Comment by atruskie** [comment]
I'm still new to AASM but I have just been bitten by `to_state` being ignored in this code [event.rb#L124](https://github.com/aasm/aasm/blob/0dd67f654343fd39498388e8e4615ceb2dadea0c/lib/aasm/core/event.rb#L124). 

Similar to @hugueslamy, the first argument I passed to the event (`to_state`) was not  a valid to state for the event and the first transition succeeded. I expected an `InvalidTransition` error.

I personally find value in the idea of being able to pass the desired state into an event extremely useful so that web services can patch the state but I think that's less important than confusing behaviour.

**Comment by maqicheng** [comment]
aa

**Comment by robbie-mcfarlane** [comment]
Just for anyone else coming across this, you can also use an 'after' block with arguments:

```
event :authorize do
  transitions :from => :initialized, :to => :authorized
  after do |args|
    # use args[:user] here
  end
end
```
`obj.authorize!(user: user)`
